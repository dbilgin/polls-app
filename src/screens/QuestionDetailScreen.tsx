import React from 'react';
import {
  StyleSheet,
  View,
  Text,
  Alert,
  ScrollView,
  TouchableOpacity,
} from 'react-native';
import {useSelector, useDispatch} from 'react-redux';
import {RootState} from '../store/reducers';
import {RootStackParamList} from '../../App';
import {StackNavigationProp} from '@react-navigation/stack';
import {RouteProp} from '@react-navigation/native';
import ErrorView from '../components/ErrorView';
import {sum} from '../utils';
import {voteQuestion} from '../../PollsService';
import {updateVoteWithUrl} from '../store/actions';

type ProfileScreenRouteProp = RouteProp<
  RootStackParamList,
  'QuestionDetailScreen'
>;

type QuestionDetailScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'QuestionDetailScreen'
>;

type Props = {
  route: ProfileScreenRouteProp;
  navigation: QuestionDetailScreenNavigationProp;
};

const QuestionDetailScreen: React.FC<Props> = ({route, navigation}) => {
  const dispatch = useDispatch();

  const {questionId} = route.params;
  const question = useSelector((state: RootState) =>
    state.questions.questions.find((x) => x.id === questionId),
  );

  const vote = async (choiceId: number, choiceUrl: string) => {
    if (choiceId && !!question?.id) {
      try {
        const voteResult = await voteQuestion(questionId, choiceId);
        if (voteResult.status === 201) {
          dispatch(updateVoteWithUrl(question.id, choiceUrl));
          navigation.goBack();
        } else {
          Alert.alert('An error has occurred.');
        }
      } catch {
        Alert.alert('An error has occurred.');
      }
    } else {
      Alert.alert('An error has occurred.');
    }
  };

  return !question ? (
    <ErrorView />
  ) : (
    <View style={styles.container}>
      <ScrollView>
        <View style={{padding: 24}}>
          <Text style={styles.title}>{question.question}</Text>
          <Text style={styles.subtitle}>{question.url}</Text>
          <Text style={styles.dateText}>
            Published at: {new Date(question.published_at).toLocaleDateString()}
          </Text>
          <Text style={styles.dateText}>{`This question has a total of ${sum(
            question.choices.map((x) => x.votes),
          )} votes.`}</Text>

          {question.choices.map((choice) => (
            <TouchableOpacity
              key={choice.url}
              onPress={() =>
                vote(Number(choice.url.split('/').pop()), choice.url)
              }
              style={styles.cardContainer}>
              <Text style={styles.choiceText}>{choice.choice}</Text>
            </TouchableOpacity>
          ))}
        </View>
      </ScrollView>
    </View>
  );
};

export default QuestionDetailScreen;

const styles = StyleSheet.create({
  choiceText: {
    fontSize: 18,
  },
  container: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: 'white',
  },
  bottomText: {
    fontSize: 24,
    color: 'black',
    marginTop: 24,
  },
  dateText: {
    fontSize: 16,
    color: '#505b7d',
    marginTop: 12,
  },
  subtitle: {
    fontSize: 12,
    color: '#505b7d',
    marginTop: 12,
  },
  title: {
    fontSize: 24,
    color: '#505b7d',
  },

  cardContainer: {
    margin: 18,
    height: 100,
    borderWidth: 0,
    borderRadius: 18,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'white',

    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.3,
    shadowRadius: 4.65,

    elevation: 8,
  },
});
