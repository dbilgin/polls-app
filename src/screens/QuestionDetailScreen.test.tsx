import React from 'react';
import {shallow} from 'enzyme';
import {TouchableOpacity} from 'react-native';
import QuestionDetailScreen from './QuestionDetailScreen';

const createTestProps = () => ({
  route: {
    params: {
      questionId: 1,
    },
  },
  navigation: {
    navigate: jest.fn(),
    goBack: jest.fn(),
  },
});

jest.mock('../PollsService', () => ({
  voteQuestion: jest.fn(() => Promise.resolve({status: 201})),
}));

jest.mock('react-redux', () => ({
  useSelector: () => {
    return {
      id: 1,
      question: 'question',
      choices: [{choice: 'choice 1', url: 'url/1', votes: 0}],
      published_at: new Date(),
      url: '/url/1',
    };
  },
  useDispatch: () => jest.fn(),
}));

describe('QuestionAddScreen', () => {
  describe('Rendering', () => {
    let props: any;
    beforeEach(() => {
      props = createTestProps();
    });

    it('should match to snapshot', () => {
      const component = shallow(<QuestionDetailScreen {...props} />);
      expect(component).toMatchSnapshot();
    });
  });
});

describe('Interaction', () => {
  let props: any;
  beforeEach(() => {
    props = createTestProps();
  });

  it('should vote and go back', async () => {
    const component = shallow(<QuestionDetailScreen {...props} />);

    // @ts-ignore
    await component.find(TouchableOpacity).first().props().onPress();
    expect(props.navigation.goBack).toHaveBeenCalled();
  });
});
