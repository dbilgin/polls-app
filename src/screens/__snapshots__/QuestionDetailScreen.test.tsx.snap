// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionAddScreen Rendering should match to snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ScrollViewMock>
    <View
      style={
        Object {
          "padding": 24,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#505b7d",
            "fontSize": 24,
          }
        }
      >
        question
      </Text>
      <Text
        style={
          Object {
            "color": "#505b7d",
            "fontSize": 12,
            "marginTop": 12,
          }
        }
      >
        /url/1
      </Text>
      <Text
        style={
          Object {
            "color": "#505b7d",
            "fontSize": 16,
            "marginTop": 12,
          }
        }
      >
        Published at: 
        6/7/2020
      </Text>
      <Text
        style={
          Object {
            "color": "#505b7d",
            "fontSize": 16,
            "marginTop": 12,
          }
        }
      >
        This question has a total of 0 votes.
      </Text>
      <TouchableOpacity
        activeOpacity={0.2}
        key="url/1"
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderRadius": 18,
            "borderWidth": 0,
            "elevation": 8,
            "height": 100,
            "justifyContent": "center",
            "margin": 18,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 4,
              "width": 0,
            },
            "shadowOpacity": 0.3,
            "shadowRadius": 4.65,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 18,
            }
          }
        >
          choice 1
        </Text>
      </TouchableOpacity>
    </View>
  </ScrollViewMock>
</View>
`;
